Add some globalization recommendations for best practices. (#40492)
dotnet/docs: docs/core/extensions/best-practices-for-developing-world-ready-apps.md

1. If a security decision is based on the result of a string comparison or case change operation, use a culture-insensitive string operation. This practice ensures that the result is not affected by the value of `CultureInfo.CurrentCulture`. See the ["String Comparisons that Use the Current Culture"](../../standard/base-types/best-practices-strings.md#string-comparisons-that-use-the-current-culture) section of [Best Practices for Using Strings](../../standard/base-types/best-practices-strings.md) for an example that demonstrates how culture-sensitive string comparisons can produce inconsistent results.

1. For any element being used for interchange (eg, a field in a JSON document in an API call) or storage, use the <xref:System.Globalization.CultureInfo?displayProperty=InvariantCulture>; additionally, you should explicitly specify a roundtrip format (such as the [`"O"`, `"o"` date-time format specifier](../../standard/base-types/standard-date-and-time-format-strings.md#the-round-trip-o-o-format-specifier)).  Although the format strings for the invariant culture are stable and unlikely to change, specifying an explicit format string helps to clarify the intent of your code.

>     - For date/time elements, consider the advice and observations in Jon Skeet's post [_Storing UTC is not a silver bullet_](https://codeblog.jonskeet.uk/2019/03/27/storing-utc-is-not-a-silver-bullet/).

Reviewer: ```suggestion
    - For date/time elements, consider the advice and observations of Noda Time author Jon Skeet, who shares valuable insights. For more information, see [Jon Skeet: _Storing UTC is not a silver bullet_](https://codeblog.jonskeet.uk/2019/03/27/storing-utc-is-not-a-silver-bullet/).
```

> 1. Globalization data [is not stable](/archive/blogs/shawnste/culture-data-shouldnt-be-considered-stable-except-for-invariant), and you should write your application and its tests with this in mind. It's updated several times a year through host OS channels on all supported platforms. This data is typically not distributed with the runtime.

Reviewer 2: ```suggestion
1. Globalization data [is not stable](https://devblogs.microsoft.com/i18n/culture-data-shouldnt-be-considered-stable/), and you should write your application and its tests with this in mind. It's updated several times a year through host OS channels on all supported platforms. This data is typically not distributed with the runtime.
```
This is an updated version of the same blog post from the same author.

> 1. Globalization data [is not stable](https://learn.microsoft.com/en-us/archive/blogs/shawnste/culture-data-shouldnt-be-considered-stable-except-for-invariant), and you should write your application and its tests with this in mind.  It is updated several times a year through host OS channels on all supported platforms.  This data is not maintained by or distributed with the runtime, nor (outside of legacy Windows applications) directly by the host OS.

Reviewer 2: ```suggestion
1. Globalization data [is not stable](https://learn.microsoft.com/en-us/archive/blogs/shawnste/culture-data-shouldnt-be-considered-stable-except-for-invariant), and you should write your application and its tests with this in mind.  It is updated several times a year through host OS channels on all supported platforms.  This data is typically not distributed with the runtime.
```
For example, iOS and wasm shipped the globalization data too, it has changed recently with hybrid globalization. I would not into enumerating all past and present cases where the globalization data ships with the runtime or the application.

1. Note that the uiCulture value can be set to the browser accept language.

1. For applications that are distributed, allow zero-downtime updates (eg, Azure container applications), or similar you must plan for situations where there may be multiple instances of the application with different format rules or other culture data, most relevantly timezone rules.

>     - If your application deployment includes a database, remember that the database will have its own globalization rules.  In most cases you should avoid performing any globalization-related functions in the database.

Reviewer: ```suggestion
    - If your application deployment includes a database, remember that the database will have its own globalization rules. In most cases you should avoid performing any globalization-related functions in the database.
```

>     - If your application deployment includes a client application or web frontend using client globalization resources, assume that the client resources differ from the resources available to your server.  Consider exclusively performing globalization functions on the client.

Reviewer: ```suggestion
    - If your application deployment includes a client application or web frontend using client globalization resources, assume that the client resources differ from the resources available to your server. Consider exclusively performing globalization functions on the client.
```

## Recommendations for robust testing

> 1. To make dependencies more explicit and testing potentially easier and parallelizable, you _SHOULD CONSIDER_ explicitly passing culture-relevant settings, such as `CultureInfo` parameters, to methods that perform formatting, and `TimeZoneInfo` to methods that work with dates and times.  You **MUST** use <xref:System.TimeProvider> or a similar type when retrieving the time.

Reviewer: ```suggestion
1. To make dependencies more explicit and testing potentially easier and parallelizable, you _should consider_ explicitly passing culture-relevant settings, such as `CultureInfo` parameters, to methods that perform formatting, and `TimeZoneInfo` to methods that work with dates and times. You _should also_ use <xref:System.TimeProvider> or a similar type when retrieving the time.
```

> 1. For most tests, you _SHOULD NOT_ explicitly validate the exact output of a given formatting operation or the exact offset of a timezone.  Formatting and timezone data may change at any time, and may differ between two otherwise identical instances of an operating system (and potentially different processes on the same machine).  Relying on an exact value makes tests brittle.

Reviewer: ```suggestion
1. For most tests, you _shouldn't_ explicitly validate the exact output of a given formatting operation or the exact offset of a time zone. Formatting and time zone data may change at any time and may differ between two otherwise identical instances of an operating system (and potentially different processes on the same machine). Relying on an exact value makes tests brittle.
```

>     - Generally, validating that some output was received will be sufficient (eg, non-empty strings when formatting).

Reviewer: ```suggestion
    - Generally, validating that some output was received will be sufficient (for example, non-empty strings when formatting).
```

>     - For some data elements and formats, validating that the data parses to the input value may be used instead (roundtripping).  Care needs to be taken for cases where fields are dropped (eg, year for some date-related fields) or the value truncated or rounded (such as for floating-point output).

Reviewer: ```suggestion
    - For some data elements and formats, validating that the data parses to the input value may be used instead (roundtripping). Care needs to be taken for cases where fields are dropped (for example, year for some date-related fields) or the value truncated or rounded (such as for floating-point output).
```

>     - If you have explicit requirements to validate all localized format output, you _SHOULD CONSIDER_ creating and using a custom culture during test setup.  For most simple cases this can be done by instantiating a `CultureInfo` object via `new CultureInfo(..)` and setting the `DateTimeFormat` and `NumberFormat` properties, and can be entirely stable.  For more complicated cases, subclassing the type allows overriding additional properties.  There are potential additional benefits to this, such as enabling [pseudolocalization](https://learn.microsoft.com/en-us/globalization/methodology/pseudolocalization) via resource files.

Reviewer: ```suggestion
    - If you have explicit requirements to validate all localized format output, you _should consider_ creating and using a custom culture during test setup. For most simple cases this can be done by instantiating a `CultureInfo` object with its constructor `new CultureInfo(..)` and setting the `DateTimeFormat` and `NumberFormat` properties. For more complicated cases, subclassing the type allows overriding additional properties. There are potential additional benefits to this, such as enabling [pseudolocalization](/globalization/methodology/pseudolocalization) with resource files.
```

>     - If you have explicit requirements to validate the results of all date/time operations, you _SHOULD CONSIDER_ creating and using a custom `TimeZoneInfo` instance during test setup.  There are potential additional benefits to this, such as enabling stable testing of certain edge cases (eg, changes to DST rules).

Reviewer: ```suggestion
    - If you have explicit requirements to validate the results of all date/time operations, you _should consider_ creating and using a custom `TimeZoneInfo` instance during test setup. There are potential additional benefits to this, such as enabling stable testing of certain edge cases (for example, changes to DST rules).
```
